src/order/generic_cofilter.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2020 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import order.basic
+import data.equiv.encodable
+
+/-!
+# Generic cofilters
+
+This file proves the dual Rasiowa–Sikorski lemma: given a preorder `P`, a term `p : P`, and
+a countable family `𝒟` of cofinal subsets of `P`, we construct a downwards closed, upwards
+directed subset of `P` which contains `p` and intersects every set in `𝒟`.
+
+## Main definitions
+
+- `generic.cofilter p 𝒟`: the constructed subset of `P`, a `𝒟`-generic cofilter.
+- `generic.directed_on p 𝒟`: the fact that the generic cofilter is upwards directed.
+- `generic.meets p 𝒟`: the fact that the generic cofilter intersects all sets in `𝒟`.
+
+## Usage
+
+This provides a framework for certain recursive constructions, similar to Zorn's lemma.
+Loosely speaking, suppose we want to construct an object satisfying some countable family
+of conditions. In this framework, we need to:
+
+- Define a type `P`, whose terms should represent finitary attempts at constructing the object.
+- Provide a 'starting point' `p : P` for the construction.
+- Define a reflexive and transitive order `≤` on `P`, where `x ≤ y` should mean '`y` extends `x`'.
+- For each condition, define a subset of `P`, consisting of attemps which guarantee the condition,
+  together with a proof that any attempt can be extended so as to guarantee the condition.
+- Use `generic.cofilter` to define the desired object, as a limit of the attempts.
+
+## References
+
+https://en.wikipedia.org/wiki/Rasiowa–Sikorski_lemma
+
+## Tags
+
+generic, filter, countable, cofinal, dense
+
+-/
+
+variables {P : Type*} [preorder P]
+
+/-- A downwards closed set, or initial segment. -/
+def downwards_closed (s : set P) : Prop := ∀ (x : P) (y ∈ s), x ≤ y → x ∈ s
+
+/-- A subset of a preorder is cofinal if it contains arbitrarily large elements. -/
+def cofinal (s : set P) : Prop := ∀ x : P, ∃ y ∈ s, x ≤ y
```

Comment:
These two belong elsewhere, I think - they're certainly useful outside of generic cofilters